core/client/mutationclient/client.go

```diff
@@ -37,74 +38,79 @@ const pageSize = 16
 
 // Client queries the server side mutations API.
 type Client struct {
-	client     pb.MutationServiceClient
+	client     pb.KeyTransparencyServiceClient
 	pollPeriod time.Duration
 }
 
+// EpochMutations contains all the mutations needed to advance
+// the map from epoch-1 to epoch.
+type EpochMutations struct {
+	Epoch     *pb.Epoch
+	Mutations []*pb.MutationProof
+}
+
 // New initializes a new mutations API monitoring client.
-func New(client pb.MutationServiceClient, pollPeriod time.Duration) *Client {
+func New(client pb.KeyTransparencyServiceClient, pollPeriod time.Duration) *Client {
 	return &Client{
 		client:     client,
 		pollPeriod: pollPeriod,
 	}
 }
 
-// StartPolling initiates polling the server side mutations API. It does not
-// block returns a channel.
-// The caller should listen on the channel to receiving the latest polled
-// mutations response including all paged mutations. If anything went wrong
-// while polling the response channel contains an error.
-func (c *Client) StartPolling(domainID string, startEpoch int64) (<-chan *pb.GetMutationsResponse, <-chan error) {
-	response := make(chan *pb.GetMutationsResponse)
-	errChan := make(chan error)
-	go func() {
-		epoch := startEpoch
-		t := time.NewTicker(c.pollPeriod)
-		for now := range t.C {
-			glog.Infof("Polling: %v", now)
-			// time out if we exceed the poll period:
-			ctx, cancel := context.WithTimeout(context.Background(), c.pollPeriod)
-			monitorResp, err := c.PollMutations(ctx, domainID, epoch)
-			cancel()
-			if err != nil {
-				glog.Infof("PollMutations(_): %v", err)
-				errChan <- err
-			} else {
-				// only write to results channel and increment epoch
-				// if there was no error:
-				glog.Infof("Got response %v at %v", monitorResp.Epoch, now)
-				response <- monitorResp
-				epoch++
+// StreamEpochs repeatedly fetches epochs and sends them to out until GetEpoch
+// returns an error other than NotFound or until ctx.Done is closed.  When
+// GetEpoch returns NotFound, it waits one pollPeriod before trying again.
+func (c *Client) StreamEpochs(ctx context.Context, domainID string, startEpoch int64, out chan<- *pb.Epoch) error {
+	wait := time.NewTicker(c.pollPeriod).C
+	for i := startEpoch; ; {
+		// time out if we exceed the poll period:
+		cctx, cancel := context.WithTimeout(ctx, c.pollPeriod)
```

Comment:
What do you think about scoping cctx within a local func?
```
epoch, err := func() {
  ctx, cancel := context.WithTimeout(ctx, c.pollPeriod)
  defer cancel()
  return c.client.GetEpoch(ctx, ... )
}()
```
So the derived context doesn't live longer than necessary and cancel is done with defer which keeps it close to where it is obtained?